File: apps/spark/java/sparker-server/src/main/java/com/cloudera/hue/sparker/server/sessions/SessionManager.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.cloudera.hue.sparker.server.sessions;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.util.Enumeration;
import java.util.UUID;
import java.util.concurrent.*;

public class SessionManager {

    private static final Logger LOG = LoggerFactory.getLogger(SparkSession.class);

    public enum SessionType {
        SCALA,
        PYTHON,
    }

    private ConcurrentHashMap<String, Session> sessions = new ConcurrentHashMap<String, Session>();
    private BlockingQueue<Session> freshScalaSessions = new LinkedBlockingQueue<Session>(5);

    SessionManagerGarbageCollector gcThread = new SessionManagerGarbageCollector();
    SessionCreator creatorThread = new SessionCreator(SessionType.SCALA);

    public SessionManager() {
        gcThread.setDaemon(true);
        gcThread.start();

        creatorThread.setDaemon(true);
        creatorThread.start();
    }

    public Session get(String id) throws SessionNotFound {
        Session session = sessions.get(id);
        if (session == null) {
            throw new SessionNotFound(id);
        }
        return session;
    }

    public Session create(SessionType type) throws IllegalArgumentException, IOException, InterruptedException {
        Session session;
        switch (type) {
            case SCALA: session = freshScalaSessions.take(); break;
            //case PYTHON: session = new PySparkSession(id); break;
            default: throw new IllegalArgumentException("Invalid language specified for shell session");
        }
        sessions.put(session.getId(), session);
        return session;
    }

    public void close() throws InterruptedException, IOException, TimeoutException {
        for (Session session : sessions.values()) {
            sessions.remove(session.getId());
            session.close();
        }

        gcThread.interrupt();
        gcThread.join();
        creatorThread.interrupt();
        creatorThread.join();

        Session session;
        while ((session = freshScalaSessions.poll(500, TimeUnit.MILLISECONDS)) != null) {
            session.close();
        }
    }

    public void close(String id) throws InterruptedException, TimeoutException, IOException, SessionNotFound {
        Session session = this.get(id);
        sessions.remove(id);
        session.close();

    }

    public Enumeration<String> getSessionIds() {
        return sessions.keys();
    }

    public void garbageCollect() throws InterruptedException, IOException, TimeoutException {
        long timeout = 60000; // Time in milliseconds; TODO: make configurable
        for (Session session : sessions.values()) {
            long now = System.currentTimeMillis();
            if ((now - session.getLastActivity()) > timeout) {
                try {
                    this.close(session.getId());
                } catch (SessionNotFound sessionNotFound) {
                    // Ignore
                }
            }
        }
    }

    private class SessionManagerGarbageCollector extends Thread {

        protected long period = 1000 * 60 * 60; // Time in milliseconds; TODO: make configurable

        public SessionManagerGarbageCollector() {
            super();
        }

        public void run() {
            try {
                while(true) {
                    garbageCollect();
                    sleep(period);
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            } catch (TimeoutException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public class SessionNotFound extends Throwable {
        public SessionNotFound(String id) {
            super(id);
        }
    }

    private class SessionCreator extends Thread {
        SessionType type;

        public SessionCreator(SessionType type) {
            this.type = type;
        }

        public void run() {
            try {
                while(true) {
                    String id = UUID.randomUUID().toString();

                    Session session;
                    switch (type) {
                        case SCALA: session = new SparkSession(id); break;
                        //case PYTHON: session = new PythonSession(id); break;
                        default: throw new IllegalArgumentException("Invalid language specified for shell session");
                    }

                    freshScalaSessions.put(session);
                }
            } catch (InterruptedException e) {
                e.printStackTrace();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }
}

File: apps/spark/java/livy-core/src/main/scala/com/cloudera/hue/livy/sessions/SessionManager.scala
/*
 * Licensed to Cloudera, Inc. under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  Cloudera, Inc. licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.cloudera.hue.livy.sessions

import java.util.concurrent.atomic.AtomicInteger

import com.cloudera.hue.livy.{LivyConf, Logging}
import org.json4s.JValue

import scala.collection.mutable
import scala.concurrent.{ExecutionContext, Future}

object SessionManager {
  val SESSION_TIMEOUT = "livy.server.session.timeout"
}

class SessionManager[S <: Session](livyConf: LivyConf, factory: SessionFactory[S])
  extends Logging {

  private implicit def executor: ExecutionContext = ExecutionContext.global

  private[this] final val _idCounter = new AtomicInteger()
  private[this] final val _sessions = mutable.Map[Int, S]()

  private[this] final val sessionTimeout = livyConf.getInt(SessionManager.SESSION_TIMEOUT, 1000 * 60 * 60)
  private[this] final val garbageCollector = new GarbageCollector

  garbageCollector.setDaemon(true)
  garbageCollector.start()

  def create(createRequest: JValue): S = {
    val id = _idCounter.getAndIncrement
    val session: S = factory.create(id, createRequest)

    info("created session %s" format session.id)

    synchronized {
      _sessions.put(session.id, session)
      session
    }
  }

  def get(id: Int): Option[S] = _sessions.get(id)

  def size(): Int = _sessions.size

  def all(): Iterable[S] = _sessions.values

  def delete(id: Int): Option[Future[Unit]] = {
    get(id).map(delete)
  }

  def delete(session: S): Future[Unit] = {
    session.stop().map { case _ =>
      synchronized {
        _sessions.remove(session.id)
      }

      Unit
    }
  }

  def shutdown(): Unit = {}

  def collectGarbage(): Future[Iterable[Unit]] = {
    def expired(session: Session): Boolean = {
      session.lastActivity.orElse(session.stoppedTime) match {
        case Some(lastActivity) =>
          val currentTime = System.currentTimeMillis()
          currentTime - lastActivity > sessionTimeout
        case None =>
          false
      }
    }

    Future.sequence(all().filter(expired).map(delete))
  }

  private class GarbageCollector extends Thread("session gc thread") {

    private var finished = false

    override def run(): Unit = {
      while (!finished) {
        collectGarbage()
        Thread.sleep(60 * 1000)
      }
    }

    def shutdown(): Unit = {
      finished = true
    }
  }
}
